[TS-4155] fix/sanitize-call-args

dvliman opened the pull request:
sanitize only `gcache_auth` call data - been told to do this on case by case basis

more details as to why I pick this solution:
1) so at first, extending error logging might sound like the right solution: http://erlang.org/doc/apps/sasl/error_logging.html since erlang (sasl) seems to print fun args in error_info / reason
2) while looking into it, I see lager attach their default error logging handler 
`(ttserver@127.0.0.1)9> gen_event:which_handlers(error_logger).`
`[error_logger_lager_h]` 
3) so essentially to strip out private customer's data (api key and secret passed in as part of any permutation of function calls) we need to replace this clause: https://github.com/erlang-lager/lager/blob/master/src/error_logger_lager_h.erl#L293-L295
4) we would have to keep the rest of the good thing that lager brought us
5) unregister it
6) register ours
7) all just to sanitize this call data: https://github.com/tigertext/epocxy/blob/master/src/cxy_ctl.erl#L605
8) so this fix is just to address that call data 
9) call data can be any number of MFA on local and remote node via rpc..

---
not bumping up epocxy version because this PR doesn't make any API change. and we have multiple consumers of multiple dependencies

devapi is locked to `erlang_commons`: `1.1.17` https://github.com/tigertext/developer_api/blob/v4/rebar.config#L17

xmpp is locked to `erlang_commons`: `master` https://github.com/tigertext/xmpp/blob/v4/ttserver/rebar.config#L56

and `erlang_commons` both `master` and `1.1.17` version is locked on `tigertext/epocxy`: `1.0.1`
https://github.com/tigertext/erlang_commons/blob/master/rebar.config#L33
https://github.com/tigertext/erlang_commons/blob/1.1.17/rebar.config#L33

erlang_commons master and 1.11.17. `cxy_ctl` is identical
https://raw.githubusercontent.com/tigertext/epocxy/master/src/cxy_ctl.erl
https://raw.githubusercontent.com/tigertext/epocxy/1.0.1/src/cxy_ctl.erl

should we have erlang_commons point to tigertext/epocxy:master (that is updating less frequently) and bump up erlang_commons, and have xmpp/devapi point to latest erlang_commons?

anderswei commented on src/cxy_ctl.erl:
```diff
@@ -601,9 +601,22 @@ execute_wrapper(Mod, Fun, Args, Task_Type, Max_History, Start, Spawn_Or_Inline,
+% pattern match for specific case - trade off between
+% valuable debugging information vs. private customer's data
+sanitize({exit, {timeout, {gen_server, call, _}},
+    {creating_new_value_with, gcache_auth, {check_auth, _}}}) ->
```
so this fix is specific for gcache_auth in cxy_ctrl, instead of generic solution?

dvliman replied:
yeah, I could have matched for gen server call timeout, function clause, badmatch, and that would hide almost everything. I am not sure that is a good idea though? it will leave us with bare stacktrace like cxy_ctl execute this and that. and that is not enough of information for debugging?

TT-lwang commented on src/cxy_ctl.erl:
```diff
@@ -601,9 +601,22 @@ execute_wrapper(Mod, Fun, Args, Task_Type, Max_History, Start, Spawn_Or_Inline,
+% valuable debugging information vs. private customer's data
+sanitize({exit, {timeout, {gen_server, call, _}},
+    {creating_new_value_with, gcache_auth, {check_auth, _}}}) ->
+    gcache_auth_timeout;
```
"gcache_auth_timeout" seems to be less informative. Can we just remove credentials, e.g api key, from the log and keep other information?

nitzanharel replied:
@dvliman - yeah, I thought that this is what we wanted to do actually and not just remove the entire payload... no?

dvliman replied:
@anderswei @TT-lwang @nitzanharel updated per your comments: https://github.com/tigertext/epocxy/pull/18/commits/f90fa0c13dfb74c25132275f46efd9d7db0a1bab

nitzanharel commented:
@dvliman - wouldn't it make more sense to change the API of cxy_ctl to accept a "santize" function as a parameter and generalize this handling externally instead of having the implementation in epocxy ?

dvliman commented:
@nitzanharel I am not sure I understand. Code example?

nitzanharel commented:
@dvliman - what I mean is to call cxy_ctl (eventually) with an optional sanitize function (default would not sanitize). that way the sanitize function logic in not in epocxy and is dynamic.

anderswei commented on src/cxy_ctl.erl:
```diff
@@ -601,9 +601,26 @@ execute_wrapper(Mod, Fun, Args, Task_Type, Max_History, Start, Spawn_Or_Inline,
+
+    {exit, {timeout, {gen_server, call, [ServerRef, Request, Timeout]}},
+        {creating_new_value_with, gcache_auth,
+            {check_auth, {api_key, xxx, xxx, CxyCtlOpts}}}};
```
can we please keep some part of the API key? E.g. several letters? That would help us debug which API has problem,

dvliman replied:
@anderswei @nitzanharel several letters might not be useful. You can be mistaken with other api keys. Can we print api key but no secret? What is the rules?

Also, the core problem here is redis connection / wpool that holds the conn goes away / timeout. We better spend time to look deeper.

Dict_Props might do but requires changes at call sites, do it later? do we not want to sanitize it?

nitzanharel replied:
The problem here is not always Redis. A timeout here can be because of endless reasons. Even if Redis is the problem, it might be because of things we can't control (e.g. network issues) - and we would still want to NOT show these things in the logs.

yes - it will require changes at the call side, but : 
1. it might end up being just one place because of how we reuse code
2. the sanitize code suggested have nothing to do with the library (epocxy) but are specific to usage
3. as we find more clauses - we will want to update this stuff and updating a library for that feels out of place

nitzanharel replied:
As for the rule - I think that printing the api key without the password should be fine.

anderswei commented:
@nitzanharel  I would rather enhance the sanitize function to add more clauses ..

nitzanharel commented:
@anderswei - we are addressing just the things we found at the moment. No one said we can't/won't enhance with more clauses. We if know of clauses - great, but I'm not sure we know of any other right now.

anderswei commented:
@nitzanharel  Passing a fun is a bad idea because that solution is trying to solve the problem in the wrong layer. For cxy_ctrl it more likes a framework and the framework should handle the logging well, instead of asking application side to handle those basic functions. 
That is why in the first place I asked David why it is not a generic solution. I will prefer a global plugin for our logging system. And there should be  some parameters to configure the plugin,  Once the plugin is active, all crash data will go through that plugin and the plugin should handle all of those requirements there well.

nitzanharel commented:
@anderswei - fair enough, only if it makes sense to go that route. Any concrete suggestions ?

anderswei commented:
@nitzanharel  as David mentioned above, there is already a plugin called error_logger_lager_h provided by lager, I would prefer to add handlers there, and pattern match sensitive data there and strip them. At first we can only filter currently needed api key but we can easily extend it and add more patterns there. Of course that function needs to walk through the call parameters and pattern match the desired pattern by several tuple_to_list and list_to_tuple. @dvliman  what do you think?

nitzanharel commented:
@anderswei - I think @dvliman and I discussed the handler and if I recall correctly, that route was quite horrible (I don't remember the details of why at the moment. Maybe David can recollect and add more color).
The other thing is that this handler in this case will go through each and every log we have which is quite annoying where in this case its really targeted to the exception part. Pros and cons for each alternative.

dvliman commented:
@anderswei, yes and no, and here is the reason why
1) yes, we can add handler (we can add multiple handlers to gen_event)
2) but we need to unregister the current `error_logger_lager_h `, more specifically, we want to override this clause: https://github.com/erlang-lager/lager/blob/master/src/error_logger_lager_h.erl#L293-L295 and keep the rest of the functionality (too good to be thrown away for nothing)
3) we can also copy paste the module and just modify that part (in theory, but it feels hacky and we will have to maintain this in the future)

all those reasons lead me to just sanitize the part we know that generates it

anderswei commented:
@nitzanharel  no we don't need to go through every line of logs. we just need to go through the "error logs" which are reported by error_logger module. by default error_logger only reports log when this is a crash.

anderswei commented:
just had a discussion with David.
the concern @dvliman  has is that the maintenance cost for a forked lager. but I think it is fine. 
anyway for a short & quick fix, this fix is good enough but if we want an extendable fix, i still prefer we go to the lager one.

nitzanharel commented:
I actually completely disagree that its fine to bare the cost of a forked lager. These things tend to explode one day in your face. As for the actual fix - I think that if we need to go into epocxy, it should be in a generic way and not a custom one (for the same reasons BTW).

dvliman commented:
Added commit to keep the apikey and xxx out the secret. Let me know if there is other enhancement.

dvliman commented:
Btw I looked into the Dict_props suggestion. It looks like the dict props wasn't build for this purpose. For the inline case, it sets it to `[]` https://github.com/dvliman/epocxy/blob/111acf3d8cbee730a23dc3ec284fd0c52f07344c/src/cxy_ctl.erl#L410-L411 and not only that it set Dict props into the calling process dictionary: https://github.com/dvliman/epocxy/blob/111acf3d8cbee730a23dc3ec284fd0c52f07344c/src/cxy_ctl.erl#L400, https://github.com/dvliman/epocxy/blob/111acf3d8cbee730a23dc3ec284fd0c52f07344c/src/cxy_ctl.erl#L555

nitzanharel commented:
not sure why you say its not built for this purpose. its used to pass "stuff". 
When invoked it can have : 
- none -> don't send anything
- all_keys -> pass all the process dictionary of the calling process (because if you spawn and the MFA you call rely on state in the process dictionary, you want it in the spawned process)
- custom list -> send whatever you need

The first line you sent sends a [] but the dictionary is initialized one line before.
The other lines are getting the dictionary from the calling process so that they can be set in the spawned one.

I'm not sure where the problem is....

dvliman commented:
so here is the fix you are proposing:
1. because it is a gcache_auth, the replenish fun is here: https://github.com/tigertext/developer_api/blob/v4/apps/tt_devapi/src/gcache_auth.erl#L14
2. now devapi connection routes to xmpp: https://github.com/tigertext/developer_api/blob/v4/apps/tt_devapi/src/tt_devapi_connection.erl#L511-L512
3. and it gets to this: https://github.com/tigertext/xmpp/blob/v4/ttserver/apps/ttserver/src/devapi/devapi_router.erl#L47
4. and presumably redis crash / timeout happens on `tt_api:get_api` or any call really
5. which calls to: https://github.com/tigertext/xmpp/blob/v4/ttserver/apps/ttserver/src/tt_api.erl#L103
6. and it calls to: https://github.com/tigertext/xmpp/blob/v4/ttserver/apps/ttserver/src/tt_support.erl#L483
7. and here you would pass in a tuple list containing a function with arity 1, taking call data, and you pass in this function if the config flag say so
8. and if it is inline: https://github.com/tigertext/epocxy/blob/master/src/cxy_ctl.erl#L411 you still pass in that function, so you may want to sanitize on Call_Data?

nitzanharel commented:
@dvliman - thanks for the detailed info.
I'm assuming you tracked the call and therefore I'm assuming the information you provided is correct.
If that is the case, then - yes - in the call to "cxy_ctl:execute_pid_monitor" we would send another parameter which is the dictionary.
But for line item 8, we would not sanitize the data there but rather, in the place where you sanitized the data in the original PR - in the "fail_wrapper" - check if there is a sanitize function present and if so - apply it on the data (otherwise just pass it as is).
So there is still code to change in epocxy but it would be a generic change and not too specific to our implementation or to that specific code path. 
Does that make sense?

dvliman commented:
Sure

no, for item 8, I am not sanitizing there - I mean I would need to change the very last `[]` to carry along dict-props with sanitize function (and hopefully it doesn't break anything else)

I'd also need to update here: https://github.com/tigertext/developer_api/blob/v4/apps/tt_devapi/src/tt_devapi_connection.erl#L532
and everywhere else that calls `cxy_ctl:execute_pid_monitor` or any other call sites to cxy_ctl public method.

nitzanharel commented:
seems about right. The only thing is that I think that https://github.com/tigertext/developer_api/blob/v4/apps/tt_devapi/src/tt_devapi_connection.erl#L532 is the actual place you need to change. the tt_support might not be relevant for this ticket as this ticket is about the specific stack trace in question (I might be wrong).
With that said - the tt_support might send a function that is different based on the different caches it is using (a.k.a - don't always send one function that takes care of all the things in the world if possible but try to use the right tool (right function) in the right place and time)

dvliman commented:
closing this PR, addressed all comments on these 2 PR:
https://github.com/tigertext/epocxy/pull/19/files
https://github.com/tigertext/developer_api/pull/2066

passing sanitize fn at caller site and apply it at epocxy if any
